src/Integration/Binding/ProjectRuleSetWriter.cs

```diff
@@ -1,235 +1,248 @@
+            : base(fs)
+        {
+        }
+
+        /// <summary>
+        /// Write a project-level SonarQube <see cref="RuleSet"/> file in the <param name="projectRoot"/> directory.
+        /// </summary>
+        /// <param name="projectFullPath">The absolute full path to the project</param>
+        /// <param name="ruleSetFileName">The rule set file name</param>
+        /// <param name="solutionRuleSetPath">Full path of the parent solution-level SonarQube rule set</param>
+        /// <param name="existingRuleSetPath">Existing project rule set</param>
+        /// <returns>Full file path of the file that was written out</returns>
+        public string WriteProjectLevelRuleSet(string projectFullPath, string ruleSetFileName, string solutionRuleSetPath, string currentRuleSetPath)
+        {
+            if (string.IsNullOrWhiteSpace(projectFullPath))
+            {
+                throw new ArgumentNullException(nameof(projectFullPath));
+            }
+
+            if (string.IsNullOrWhiteSpace(ruleSetFileName))
+            {
+                throw new ArgumentNullException(nameof(ruleSetFileName));
+            }
+
+            if (string.IsNullOrWhiteSpace(solutionRuleSetPath))
+            {
+                throw new ArgumentNullException(nameof(solutionRuleSetPath));
+            }
+
+            string projectRoot = Path.GetDirectoryName(projectFullPath);
+            string ruleSetRoot = PathHelper.ForceDirectoryEnding(projectRoot);
+
+            string existingRuleSetPath;
+            if (this.TryUpdateExistingProjectRuleSet(solutionRuleSetPath, ruleSetRoot, currentRuleSetPath, out existingRuleSetPath))
+            {
+                Debug.Assert(existingRuleSetPath != null);
+                return existingRuleSetPath;
+            }
+
+            // Create a new project level rule set
+            string solutionIncludePath = PathHelper.CalculateRelativePath(ruleSetRoot, solutionRuleSetPath);
+            RuleSet newRuleSet = GenerateNewProjectRuleSet(solutionIncludePath, currentRuleSetPath);
+            string newRuleSetPath = this.GenerateNewProjectRuleSetPath(ruleSetRoot, ruleSetFileName);
+            this.FileSystem.WriteRuleSetFile(newRuleSet, newRuleSetPath);
+
+            return newRuleSetPath;
+        }
+
+        #region Rule Set Helpers
+        internal /*for testing purposes*/ bool TryUpdateExistingProjectRuleSet(string solutionRuleSetPath, string projectRuleSetRootFolder, string currentRuleSet, out string existingRuleSetPath)
+        {
+            existingRuleSetPath = null;
+            if (ShouldIgnoreConfigureRuleSetValue(currentRuleSet))
+            {
+                return false;
+            }
+
+            existingRuleSetPath = PathHelper.ResolveRelativePath(currentRuleSet, projectRuleSetRootFolder);
+            if (!PathHelper.IsPathRootedUnderRoot(existingRuleSetPath, projectRuleSetRootFolder))
+            {
+                // Not our file (i.e. absolute path to some other ruleset)

```

Comment:
Oh, sorry - missed that. It's really hard to see the actual diff on GitHub because pretty much every file is being shown as every line has changed, when it hasn't